libs/libc/stdio/lib_fmemopen.c

```diff
@@ -0,0 +1,291 @@
+
+  fmemopen_cookie->pos += size;
+  if (fmemopen_cookie->pos > fmemopen_cookie->end)
+    {
+      fmemopen_cookie->end = fmemopen_cookie->pos;
+    }
+
+  /* POSIX states that NULL byte shall be written at the current position
+   * or end of the buffer.
+   */
+
+  if (fmemopen_cookie->pos < fmemopen_cookie->size &&
+      fmemopen_cookie->buf[fmemopen_cookie->pos - 1] != '\0')
+    {
+      fmemopen_cookie->buf[fmemopen_cookie->pos] = '\0';
+    }
+
+  return size;
+}
+
+/****************************************************************************
+ * Name: fmemopen_seek
+ ****************************************************************************/
+
+static off_t fmemopen_seek(FAR void *c, FAR off_t *offset, int whence)
+{
+  FAR struct fmemopen_cookie_s *fmemopen_cookie =
+    (FAR struct fmemopen_cookie_s *)c;
+  off_t new_offset;
+
+  switch (whence)
+    {
+      case SEEK_SET:
+        new_offset = *offset;
+        break;
+      case SEEK_END:
+        new_offset = fmemopen_cookie->end + *offset;
+        break;
+      case SEEK_CUR:
+        new_offset = fmemopen_cookie->pos + *offset;
+        break;
+      default:
+        return ENOTSUP;
+    }
+
+  /* Seek to negative value or value larger than maximum size shall fail. */
+
+  if (new_offset < 0 || (new_offset > fmemopen_cookie->end))
+    {
+      return EINVAL;
+    }
+
+  fmemopen_cookie->pos = new_offset;
+  *offset = new_offset;
+  return new_offset;
+}
+
+/****************************************************************************
+ * Name: fmemopen_close
+ ****************************************************************************/
+
+static int fmemopen_close(FAR void *c)
+{
+  FAR struct fmemopen_cookie_s *fmemopen_cookie =
+    (FAR struct fmemopen_cookie_s *)c;
+  if (fmemopen_cookie->custom)
+    {
+      lib_free(fmemopen_cookie->buf);
+    }
+
+  lib_free(fmemopen_cookie);
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: fmemopen
+ ****************************************************************************/
+
+FAR FILE *fmemopen(FAR void *buf, size_t size, FAR const char *mode)
+{
+  cookie_io_functions_t fmemopen_io;
+  FAR struct fmemopen_cookie_s *fmemopen_cookie;
+  FAR FILE *filep;
+
+  fmemopen_cookie = lib_zalloc(sizeof(struct fmemopen_cookie_s));
+  if (fmemopen_cookie == NULL)
+    {
+      set_errno(ENOMEM);
+      return NULL;
+    }
+
+  if (buf == NULL)
+    {
+      /* POSIX standard states:
+       * Because this feature is only useful when the stream is opened for
+       * updating (because there is no way to get a pointer to the buffer)
+       * the fmemopen() call may fail if the mode argument does not
+       * include a '+'.
+       */
+
+      if (mode[1] != '+')
```

Comment:
let's call lib_mode2oflags to parse the flags for us, or call fopencookie first and then check filep->fs_flags instead

Reply:
There is a difference in modes between `fopen` and `fmemopen`. `fopen` truncates the buffer when `w` mode is used, `fmemopen` truncates it only when `w+` is used (see https://pubs.opengroup.org/onlinepubs/9699919799/functions/fmemopen.html). So I don't think I can use `lib_mode2oflags` or `fopencookie` as it would incorrectly set `O_TRUNC`.

Reply:
Yes, but since fdopen/lib_mode2oflags doesn't do any real action, we can reuse the parser result, but do the action a little bit differently.

Reply:
Yes but it would return me `O_TRUNC` for `w` which is not correct in case of `fmemopen`. So I would still have to check

```
if (mode[0] == 'w')
```

I suppose I could use `lib_mode2oflags` and then check `mode[0] == 'w'` in case it returns `O_TRUNC`. That would simplify the checks and let most of the work on  `lib_mode2oflags`.

Reply:
but from the link:
For modes w and w+ the initial size shall be zero.
both should truncate the stream.

Reply:
Hm, ok, seems I have misread that. Updated.